keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, there are many issues that exist during the node drain phase. This KEP analyzes the
+issues and proposes a new API for node maintenance. This API can be used to implement additional
+capabilities concerning node drain.
+
+## Motivation
+The goal of this KEP is to analyze and improve node maintenance in Kubernetes.
+
+Node maintenance is a request from a cluster administrator to remove all pods
+from a node, so that it can be disconnected from the cluster to perform a 
+software upgrade (OS, kubelet), hardware upgrade, or simply to remove the node
+as it is no longer needed.
+
+Currently, Kubernetes supports this use case in the following way with `kubectl drain`:
+1. There are running pods on node A, some of which are protected with PodDisruptionBudgets (PDB).
+2. Set the node `Unschedulable` (cordon) to prevent new pods from being scheduled there.
+3. Evict (default behavior) pods from node A by using the eviction API (see [kubectl drain worklflow](https://kubernetes.io/images/docs/kubectl_drain.svg)).
+4. Proceed with the maintenance and shut down the node.
+5. Kubelet can try to delay the shutdown to allow the remaining pods to terminate gracefully
+   ([graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown)). 
+   It can also take the priority class into an account ([pod-priority-graceful-node-shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#pod-priority-graceful-node-shutdown))
+
+The main problem is that the current approach tries to solve this in an application agnostic way
+and just tries to get rid of all the pods on the node. Since this approach cannot be applied
+generically to all pods, we define special [drain filters](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/kubectl/pkg/drain/filters.go#L153-L162) 
+that either skip groups of pods or an admin has to consent to override those groups to be either
+skipped or deleted. This means that without knowledge of all the underlying applications on the
+cluster, the admin has to make a potentially harmful decision.
+
+From an application owner or developer perspective, the only standard tool they have is
+a PodDisruptionBudget. This is sufficient in a basic scenario with a simple multi-replica
+application. The edge case applications, where this does not work are very important to
+the cluster admin, as they can block the node drain. And, in turn, very important to the
+application owner, as the admin can then override the pod disruption budget and disrupt their
+sensitive application anyway.
+
+List of cases where the current solution is not optimal:
+
+1. Single replica application have to tolerate application downtime during the node drain.
```

Comment:
```suggestion
1. Without extra manual effort, an application running with a single replica has to settle for
   experiencing application downtime during the node drain.
```

Reply:
BTW this also affects the minimum scale for HorizontalPodAutoscaler.

Reply:
I have added a new `2.` point which talks about HPAs and references the relevant issue: https://github.com/kubernetes/kubernetes/issues/93476

Reply:
Thanks for bringing this up